CI/CD Piplelines

FloezeTv opened the pull request:
This PR adds CI/CD pipelines for automatically building the different versions of the app as well as automated testing and linting.

## Changes

### Added/Modified Pipelines

- Also build web container for `arm64`
- Nighly Web-Builds to GitHub Pages
- Desktop Version: Linux, macOS, Windows on `x86_64`, `arm64`
- Build & Publish the Python wheels
- Lint rust code using `clippy`
- Check formatting using `rustfmt`
- Execute rust tests
- Publish to crates.io

### Build improvements

- Cache dependencies in `Dockerfile`

### Other changes

- Fix rust-version to `1.88`
- Fix warnings by `clippy`
- Update `egui`, `wgpu`, and `glyphon` to have `egui` depend on a released version instead of commit

Closes #90.

FloezeTv commented:
> Just a small side comment: I'd be in favor of dropping 32bit support. I have not come across such a system in the last 6+ years working on quantum computing software.

To be honest, I have only added them because they have [Tier 1-support](https://doc.rust-lang.org/nightly/rustc/platform-support.html#tier-1-with-host-tools) in rust and I thought it would not hurt to build for these platforms too.

FloezeTv commented on .github/workflows/python.yaml:
```diff
@@ -0,0 +1,115 @@
+    needs: [build, sdist]
+    environment:
+      name: pypi
+      url: https://pypi.org/p/mqt-naviz
```
Someone needs to create this package and enable trusted publishing from this repo.

burgholzer commented on .github/workflows/commitlint.yaml:
As already discussed in the other PR, please let's not lint commit messages.
The PR here already serves as the perfect example for why this is undesirable.
You needed to explain why something in the CI was red, even though that failure is completely inconsequential.
Furthermore, once CI is set up, we will set up branch protection, which will include that all CI checks must pass before merging (if not using admin privileges).
We want to be able to auto-merge PRs and not have to rely on force merges.

Please remove the respective configuration.

FloezeTv replied:
Ok, I can see that.
But I also very much dislike non-speaking commit-messages like `Minor Changes` or `Push something` which make it impossible to know what a commit does.
At that point one can just not write a commit message at all.
I think we should still try to enforce some structure in the commits, and in my opinion having to think about what to write in a commit message helps a huge step with that (e.g., thinking how to group the changes into commits).
Though I would also be fine to simply document that and taking it into account before merging a PR.

burgholzer replied:
As iterated before, I agree with you from a philosophical point of view, but can already tell you that it won't be enforceable in practice.
Given how the whole MQT defaults to squash merging anyway, it's also not as important in PRs as you make it out to be.
Our standard contributing guidelines already contain wording that advocates for writing meaningful PR titles, descriptions, etc.
That's likely to be as good as it gets (modulo updates to the wording).

FloezeTv replied:
> Given how the whole MQT defaults to squash merging anyway

I guess that would solve bad commit messages, though I currently prefer merging PRs with merge commits as it keeps the more fine-grained commits but also shows what belongs together to a single feature (though especially for single-commit-changes it does not make that much of a difference).
As an example, I think the git history would benefit from having all the separate commits in this PR instead of squashing them into a single huge commit.

As I said, I'm fine with removing the linting of commit messages but would like to still keep an eye on that when reviewing PRs.

burgholzer commented on .github/release.yaml:
For the sake of consistency, please use the release drafter configuration and workflow that is set up in all other MQT repositories.
The MQT Core one is probably the best reference here, as it also has multiple components.
You can add some details based on the text in the other workflows.

denialhaag replied:
I'm adding `release-drafter.yaml` in #102, but it may make sense that this happens here instead, especially if we merge this PR first.

FloezeTv replied:
I think it would be fine to add the release-drafter in #102 as long as it is merged before the next release.
Alternatively I can cherry-pick the relevant parts from #102.

burgholzer commented on .github/workflows/desktop-build.yaml:
Let's stick with a similar release workflow as we use it for all other MQT repositories.
That is, trigger on 
```yaml
on:
  release:
    types: [published]
  workflow_dispatch:
```

I am pretty sure that one can also directly add assets to the release that ws just published and triggered the event.

FloezeTv replied:
I can see if uploading to existing releases is supported by `softprops/action-gh-release` or another action.
From the issues there it seemed like one would need to manually use the `gh`-cli.

Do note, however, that `naviz` requires a git to display correct version information.

burgholzer commented on .github/workflows/desktop-build.yaml:
```diff
@@ -0,0 +1,110 @@
+            prepare: sudo apt install -y gcc-aarch64-linux-gnu
+            env:
+              CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_LINKER: aarch64-linux-gnu-gcc
+          - target: i686-unknown-linux-gnu
+            image: ubuntu-latest
+            prepare: sudo apt install -y gcc-i686-linux-gnu
+            env:
+              CARGO_TARGET_I686_UNKNOWN_LINUX_GNU_LINKER: i686-linux-gnu-gcc
```
```suggestion
```
Let's skip 32bit systems to safe some CI resources. They are not necessary.

burgholzer commented on .github/workflows/desktop-build.yaml:
```diff
@@ -0,0 +1,110 @@
+          - target: aarch64-pc-windows-msvc
+            image: windows-latest
+            app-suffix: .exe
+          - target: i686-pc-windows-msvc
+            image: windows-latest
+            app-suffix: .exe
```
```suggestion
```
Let's skip 32bit systems to safe some CI resources. They are not necessary.

burgholzer commented on .github/workflows/desktop-build.yaml:
```diff
@@ -0,0 +1,110 @@
+            image: macos-latest
+          - target: x86_64-apple-darwin
+            image: macos-latest
+          - target: aarch64-unknown-linux-gnu
+            image: ubuntu-latest
+            prepare: sudo apt install -y gcc-aarch64-linux-gnu
+            env:
+              CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_LINKER: aarch64-linux-gnu-gcc
```
absolutely no need to emulate here. GitHub offers native ubuntu ARM runners.

burgholzer replied:
That is interesting. I did not know that. Feel free to resolve the conversation in that case.

burgholzer commented on .github/workflows/desktop-build.yaml:
```diff
@@ -0,0 +1,110 @@
+              CARGO_TARGET_I686_UNKNOWN_LINUX_GNU_LINKER: i686-linux-gnu-gcc
+          - target: x86_64-unknown-linux-gnu
+            image: ubuntu-latest
+          - target: aarch64-pc-windows-msvc
+            image: windows-latest
+            app-suffix: .exe
```
absolutely no need to emulate here. GitHub offers native Windows ARM runners.

burgholzer reviewed (changes requested):
Thanks for your work on this @FloezeTv 🙏🏼 
For the most part, this looks great. I just have a couple of comments that should bring this a little more in line with the rest of the MQT projects and their CI, while also being mindful of CI resources in general.

FloezeTv commented on .github/workflows/publish_crates.yaml:
```diff
@@ -0,0 +1,29 @@
+          cache-key: ${{ env.CACHE_KEY }}
+      - name: Install Tools
+        run: cargo install cargo-workspaces
+      - name: Publish Crates
+        env:
+          TOKEN: ${{ secrets.CRATES_IO_TOKEN }}
+        run: cargo ws publish --token "$TOKEN" --publish-as-is --skip-published
```
After further investigating is seems that it should be available. I have previously only found [an RFC](https://rust-lang.github.io/rfcs/3691-trusted-publishing-cratesio.html), though the [accompanying tracking issue](https://github.com/rust-lang/crates.io/issues/10247) seems done.
I have not found any documentation about it, though.

FloezeTv replied:
I have found [documentation](https://crates.io/docs/trusted-publishing) and will update that workflow.

FloezeTv commented on .github/workflows/python.yaml:
```diff
@@ -0,0 +1,115 @@
+        run: |
+          echo matrix=$(jq -rc 'map([.arch=.arch[]]) | flatten' <<EOF
+            [
+              {
+                "os": "linux",
+                "image": "ubuntu-latest",
+                "arch": ["x86_64", "x86", "aarch64", "armv7", "s390x", "ppc64le"],
+                "manylinux": "auto"
+              }, {
+                "os": "musllinux",
+                "image": "ubuntu-latest",
+                "arch": ["x86_64", "x86", "aarch64", "armv7"],
+                "manylinux": "musllinux_1_2"
+              }, {
+                "os": "macos",
+                "image": "macos-latest",
+                "arch": ["x86_64", "aarch64"]
+              }, {
+                "os": "windows",
+                "image": "windows-latest",
+                "arch": ["x86_64", "x86"]
+              }
```
The target matrices are adapted from the default maturin workflow.
Again, there is no benefit from compiling on the native hardware.

FloezeTv replied:
Python-Wheel is now built for Linux and macOS `x86-64` and `aarch64` as well as Windows `x86-64`.

Regarding Windows ARM:
The windows arm image currently does not have `rustup` installed by default and ships with x86-64 versions of python which `maturin` detects and tries to link against.
We could manually install `rustup` and `aarch64` python and remove the `x86-64` python versions.
However, a new runner image with [`rustup`](https://github.com/actions/partner-runner-images/issues/77) and [native python](https://github.com/actions/partner-runner-images/issues/85) should be [released soon](https://github.com/actions/partner-runner-images/issues/77#issuecomment-3032556925).

burgholzer replied:
Then let's wait with the Windows ARM wheels until the new images are released. 👍🏼

FloezeTv commented:
I have incorporated the requested changes and rebased onto the latest `main`.
The two unresolved conversations (Windows ARM Python builds and Release drafter) are not (fully) covered by this PR and should be added in later PRs.
